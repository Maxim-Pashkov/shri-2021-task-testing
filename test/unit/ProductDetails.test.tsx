import React from 'react';
import { it, expect, describe } from '@jest/globals';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { getApplication } from './getApplication';
import { ProductDetails } from '../../src/client/components/ProductDetails';

const product = {
    name: 'test name',
    price: 10,
    count: 1,
    id: 5,
    description: 'test description',
    color: 'test color',
    material: 'test material',
};

describe('ProductDetails test', () => {
    it('item is exists', () => {
        const { application } = getApplication(
            () => <ProductDetails product={product} />,
            {}
        );

        const { container } = render(application);

        expect(container.firstChild).toMatchSnapshot();
    });
    
    it('adding multiple to cart', () => {
        const { application, store } = getApplication(
            () => <ProductDetails product={product} />,
            {}
        );

        const { queryByRole } = render(application);

        const button = queryByRole('button', { name: /add to cart/i });

        userEvent.click(button);
        userEvent.click(button);

        const { cart } = store.getState();

        expect(cart[product.id]?.count).toBe(2);
    });

    it('selected item is exists', () => {
        const cartState = {
            [product.id]: product,
        };

        const { application } = getApplication(
            () => <ProductDetails product={product} />,
            cartState
        );

        const { container } = render(application);

        expect(container.firstChild).toMatchSnapshot();
    });
});
