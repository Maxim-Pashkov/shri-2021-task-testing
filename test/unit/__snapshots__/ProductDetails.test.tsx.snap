// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails test item is exists 1`] = `
<div
  class="ProductDetails row"
>
  <div
    class="col-12 col-sm-5 col-lg-4"
  >
    <img
      class="Image"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
    />
  </div>
  <div
    class="col-12 col-sm-7 col-lg-6"
  >
    <h1
      class="ProductDetails-Name"
    >
      test name
    </h1>
    <p
      class="ProductDetails-Description"
    >
      test description
    </p>
    <p
      class="ProductDetails-Price fs-3"
    >
      $
      10
    </p>
    <p>
      <button
        class="ProductDetails-AddToCart btn btn-primary btn-lg"
      >
        Add to Cart
      </button>
    </p>
    <dl>
      <dt>
        Color
      </dt>
      <dd
        class="ProductDetails-Color text-capitalize"
      >
        test color
      </dd>
      <dt>
        Material
      </dt>
      <dd
        class="ProductDetails-Material text-capitalize"
      >
        test material
      </dd>
    </dl>
  </div>
</div>
`;

exports[`ProductDetails test selected item is exists 1`] = `
<div
  class="ProductDetails row"
>
  <div
    class="col-12 col-sm-5 col-lg-4"
  >
    <img
      class="Image"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
    />
  </div>
  <div
    class="col-12 col-sm-7 col-lg-6"
  >
    <h1
      class="ProductDetails-Name"
    >
      test name
    </h1>
    <p
      class="ProductDetails-Description"
    >
      test description
    </p>
    <p
      class="ProductDetails-Price fs-3"
    >
      $
      10
    </p>
    <p>
      <button
        class="ProductDetails-AddToCart btn btn-primary btn-lg"
      >
        Add to Cart
      </button>
      <span
        class="CartBadge text-success mx-3"
      >
        Item in cart
      </span>
    </p>
    <dl>
      <dt>
        Color
      </dt>
      <dd
        class="ProductDetails-Color text-capitalize"
      >
        test color
      </dd>
      <dt>
        Material
      </dt>
      <dd
        class="ProductDetails-Material text-capitalize"
      >
        test material
      </dd>
    </dl>
  </div>
</div>
`;
